src/domain/email/email.repository.ts

```diff
@@ -0,0 +1,53 @@
+import { IEmailDataSource } from '@/domain/interfaces/email.datasource.interface';
+import { Inject } from '@nestjs/common';
+import { ISafeRepository } from '@/domain/safe/safe.repository.interface';
+import codeGenerator from '@/domain/email/code-generator';
+import { Email } from '@/domain/email/entities/email.entity';
+import { IEmailRepository } from '@/domain/email/email.repository.interface';
+import { SignerNotOwnerError } from '@/domain/email/errors/signer-not-owner.error';
+import { EmailSaveError } from '@/domain/email/errors/email-save.error';
+
+export class EmailRepository implements IEmailRepository {
+  constructor(
+    @Inject(IEmailDataSource)
+    private readonly emailDataSource: IEmailDataSource,
+    @Inject(ISafeRepository) private readonly safeRepository: ISafeRepository,
+  ) {}
+
+  async saveEmail(args: {
+    chainId: string;
+    safeAddress: string;
+    emailAddress: string;
+    signer: string;
+  }): Promise<void> {
+    const email = new Email(args.emailAddress);
+
+    const safe = await this.safeRepository.getSafe({
+      chainId: args.chainId,
+      address: args.safeAddress,
+    });
+
+    if (!(args.signer in safe.owners)) {
+      // Signer needs to be an owner of the safe
+      throw new SignerNotOwnerError(
+        args.chainId,
+        args.safeAddress,
+        args.signer,
+      );
+    }
+
+    const verificationCode = codeGenerator();
```

Comment:
Is this PR containing the final implementation for `verificationCode`? I mean, why not simply use the already present `uuid` package to generate these verification codes instead of having a custom `codeGenerator` function?

Reply:
Not necessarily the final implementation as it will go through security review. But what is the reason to use UUID's in this case (from a security perspective given the size constraints)?

Reply:
No strong opinion here either. I was just suggesting UUID as it provides good _entrophy_, is already integrated within the project, makes the code simpler (no custom functions), and generates `string` values (which is the type `EmailDataSource.saveEmail` stores into the db and also returns).

Reply:
> I was just suggesting UUID as it provides good _entrophy_

But from a security standpoint how would we compare it with the solution that's in place 🤔 

> is already integrated within the project

`crypto` module is also part of the project or do you mean in a different way?

> makes the code simpler (no custom functions)

I'm not sure we are comparing the two functionalities in the same way. There's a single call to the `crypto` module. The remaining expressions are on the restrictions of the values (e.g.: up to 6 digits). It doesn't necessarily need to be a function that was extracted to a different file though!

> and generates string values (which is the type EmailDataSource.saveEmail stores into the db and also returns).

For this feature we will be sending digits (unless there's a security concern regarding it). The database stores it as `text` so that we remain flexible on the datasource side while moving that restriction to the domain.

Reply:
I get your points, you're right, maybe there's no reason to use something different from what it's implemented, at least before going through the security review 👍 Thanks for explaining.